common/workunit/workunit.cpp

```diff
@@ -332,113 +332,1203 @@ extern WORKUNIT_API IConstWUGraphProgress *createConstGraphProgress(const char *
+            visitor.noteStatistic(cur.getKind(), cur.getValue(), cur);
+        }
+    }
+
+    virtual bool getStat(StatisticKind kind, unsigned __int64 & value) const
+    {
+        for (unsigned i=0; i < numStatistics; i++)
+        {
+            IConstWUStatistic & cur = statistics.item(curIndex + i);
+            if (cur.getKind() == kind)
+            {
+                value = cur.getValue();
+                return true;
+            }
+        }
+        return false;
+    }
+
+    virtual const char * queryAttribute(WuAttr attr) const
+    {
+        return nullptr;
+    }
+
+    virtual const char * queryHint(const char * kind) const
+    {
+        return nullptr;
+    }
+
+protected:
+    inline IConstWUStatistic & queryStatistic(unsigned i)
+    {
+        return statistics.item(curIndex + i);
+    }
+
+    bool initScope()
+    {
+        unsigned next = curIndex+1;
+        while (next < statistics.ordinality())
+        {
+            if (compareStatisticScopes(statistics.item(curIndex), statistics.item(next)) != 0)
+                break;
+            next++;
+        }
+        numStatistics = (next - curIndex);
+        return statistics.isItem(curIndex);
+    }
+
+protected:
+    IArrayOf<IConstWUStatistic> statistics;
+    unsigned curIndex;
+    unsigned numStatistics = 1;
+};
+
+
+/*
+ * An implementation of IConstWUScopeIterator for the query graphs.
+ */
+class GraphScopeIterator : public CInterfaceOf<IConstWUScopeIterator>
+{
+private:
+    //This uses a state machine - this enumation contains the different states.
+    enum State
+    {
+        SGraphBegin,
+        SGraphFirstEdge,
+        SGraphFirstSubGraph,
+        SGraphEnd,
+        SGraphNext,
+        SSubGraphBegin,
+        SSubGraphFirstEdge,
+        SSubGraphFirstActivity,
+        SSubGraphFirstSubGraph,
+        SSubGraphEnd,
+        SSubGraphNext,
+        SEdgeBegin,
+        SEdgeNext,
+        SActivityBegin,
+        SActivityFirstSubGraph,
+        SActivityNext,
+        SDone
+    } state = SDone;
+
+public:
+    GraphScopeIterator(const IConstWorkUnit * wu, const IStatisticsFilter * _filter) : graphIter(&wu->getGraphs(GraphTypeAny))
     {
-        if (!description && createDefault)
+    }
+
+    virtual bool first() override
+    {
+        if (!graphIter->first())
+            return false;
+        state = SGraphBegin;
+        return nextScope();
+    }
+
+    virtual bool next() override
+    {
+        return nextScope();
+    }
+
+    virtual bool isValid() override
+    {
+        return graphIter->isValid();
+    }
+
+    virtual const char * queryScope() const override
+    {
+        return curScopeName.str();
+    }
+
+    virtual StatisticScopeType getScopeType() const override
+    {
+        return scopeType;
+    }
+
+    virtual void playProperties(IWuScopeVisitor & visitor) override
+    {
+        switch (scopeType)
+        {
+        case SSTgraph:
+            return;
```

Comment:
yes - because treeIters.ordinality() may be 0 for that case.  It could be removed fro the second switch.